mlos_bench/db/db_schema.sql

```diff
@@ -0,0 +1,83 @@
+--
+-- Copyright (c) Microsoft Corporation.
+-- Licensed under the MIT License.
+--
+-- DB schema for storing MLOS benchmarking results.
+-- The syntax works for SQLite3, DuckDB, PostgreSQL, and MySQL / MariaDB.
+
+DROP TABLE IF EXISTS trial_telemetry;
+DROP TABLE IF EXISTS trial_results;
+DROP TABLE IF EXISTS trial_config;
+DROP TABLE IF EXISTS experiment_merge;
+DROP TABLE IF EXISTS trial_status;
+DROP TABLE IF EXISTS experiment_config;
+
+CREATE TABLE experiment_config (
+    exp_id VARCHAR(255) NOT NULL,
+    descr TEXT,
+    git_repo VARCHAR(255) NOT NULL,
+    git_commit VARCHAR(40) NOT NULL,
+
+    PRIMARY KEY (exp_id)
+);
+
+CREATE TABLE trial_status (
+    exp_id VARCHAR(255) NOT NULL,
+    trial_id INTEGER NOT NULL,
+    ts_start TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL,
+    ts_end TIMESTAMP,
+    -- Should match the text IDs of `mlos_bench.environment.Status` enum:
+    status VARCHAR(16) NOT NULL,
+
+    PRIMARY KEY (exp_id, trial_id),
+    FOREIGN KEY (exp_id) REFERENCES experiment_config(exp_id)
+);
+
+CREATE TABLE experiment_merge (
+    dest_exp_id VARCHAR(255) NOT NULL,
+    dest_trial_id INTEGER NOT NULL,
+    source_exp_id VARCHAR(255) NOT NULL,
+    source_trial_id INTEGER NOT NULL,
+
+    UNIQUE (dest_exp_id, dest_trial_id, source_exp_id, source_trial_id),
+
+    FOREIGN KEY (dest_exp_id, dest_trial_id)
+        REFERENCES trial_status(exp_id, trial_id),
+
+    FOREIGN KEY (source_exp_id, source_trial_id)
+        REFERENCES trial_status(exp_id, trial_id)
+);
+
+CREATE TABLE trial_config (
+    exp_id VARCHAR(255) NOT NULL,
+    trial_id INTEGER NOT NULL,
+    param_id VARCHAR(255) NOT NULL,
+    param_value VARCHAR(255),
+
+    PRIMARY KEY (exp_id, trial_id, param_id),
+    FOREIGN KEY (exp_id, trial_id) REFERENCES trial_status(exp_id, trial_id),
+    FOREIGN KEY (exp_id) REFERENCES experiment_config(exp_id)
+);
+
+CREATE TABLE trial_results (
+    exp_id VARCHAR(255) NOT NULL,
+    trial_id INTEGER NOT NULL,
+    metric_id VARCHAR(255) NOT NULL,
```

Comment:
Similar to the conversation of `experiment_params` is there utility in having `experiment_metrics`?
Basically, to alleviate the experiment merge problem we want all trials with an experiment to use a similar set of "scores", right?

Reply:
We need this to store multiple metrics per experiment - one will be used as an optimization target, but we may still want to collect the others. Note that this is different from telemetry, as we store exactly one value per metric per experiment here.

Reply:
Yeah, that makes sense, but my point was if we want to be able to support reusing trials from experiments we also need to be able to know that the outputs scores from the trials are also compatible, not just the input params.
For instance, if I want to optimize for throughput in one, but I only captured time in another, then I can't reuse those trials even if they're over the same params.
Right now the only way to tell that would be to do a large distinct agg query over all the trials in an experiment.
On the other hand if the experiment logged which scores were available for all trials in it we could just check that.
Basically think of it as a materialized view or a constraint if you want.
Make sense?

Reply:
oh, I see. I thought it would be enough to fail when we try to load the data from multiple experiments into the optimizer and can't find a certain score.. 
Then let me think about it a bit more. It would probably make sense to save the id of the metric we're optimizing for along with the git info of the experiment configuration.

Reply:
As discussed previously, not all trial results need have any score (i.e. they could have been failed), but if they do have a score, I think all trials within an experiment should have the same set of score metrics, no?

Reply:
I've added an optimization target info to the experiment, and we can easily enforce that each trial (with status==SUCCEEDED) must have that field present.

Theoretically, we can also declare in advance what metrics each experiment collects and check if all of them are present when saving the data.

On the schema side, we can also enforce that we only store some known metrics for each experiment. That I understand.

What I don't understand is how we can possibly enforce - *on the schema side* - saving *all* known metrics for each trial.
Let's discuss?

Reply:
Enforcing that all possible metrics an experiment declares as exportable by the trial runner scripts I don't think is possible in the schema (though if we declared them ahead of time, we could certainly add it as a good sanity check during the `register` operation to help catch bugs in the user scripts, so sometime to add to the list for later).

Enforcing that the ones the trial does export *are* part of that set is all I was after.

Basically, it's another way to check that the experiments are merge compatible.

For instance, suppose we run two experiments `A` and `B` where we have the same set of tunable params, so that they on the surface look like they're compatible, but for experiments `B` we've changed the scripts to include metrics for both `x` and `y`, whereas `A`'s scripts only collected `x`.  If in this case we want to optimize for `x` (ignoring multi-objective tuning for a moment), then we can simply check that `x` is a declared output of all trials for experiments in `A` and `B` and proceed.  On the other hand, if `B` wants to tune for `y`, then `A`'s trials are no longer useful.
Similar for when experiment `C` comes along.

Reply:
ok I've pushed a new version of the schema that implements your `experiment -> trial -> configuration` suggestion, i.e., adds another level of indirection between the trial and its tunable parameters. I'll post an ER diagram below